src/main/java/org/spongepowered/common/item/recipe/crafting/SpongeShapedCraftingRecipeBuilder.java

```diff
@@ -0,0 +1,165 @@
+ * THE SOFTWARE.
+ */
+package org.spongepowered.common.item.recipe.crafting;
+
+import static com.google.common.base.Preconditions.checkState;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableTable;
+import com.google.common.collect.Lists;
+import it.unimi.dsi.fastutil.chars.Char2ObjectArrayMap;
+import org.spongepowered.api.item.inventory.ItemStackSnapshot;
+import org.spongepowered.api.item.recipe.crafting.ShapedCraftingRecipe;
+
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Predicate;
+
+public final class SpongeShapedCraftingRecipeBuilder implements ShapedCraftingRecipe.Builder {
+
+    private final List<String> aisle = Lists.newArrayList();
+    private final Map<Character, Predicate<ItemStackSnapshot>> ingredientMap = new Char2ObjectArrayMap<>();
+    private ItemStackSnapshot result = ItemStackSnapshot.NONE;
+
+    @Override
+    public ShapedCraftingRecipe.Builder aisle(String... aisle) {
+        this.aisle.clear();
+        ingredientMap.clear();
+
+        if (aisle != null) {
+            Collections.addAll(this.aisle, aisle);
+        }
+
+        return this;
+    }
+
+    @Override
+    public ShapedCraftingRecipe.Builder where(char symbol, Predicate<ItemStackSnapshot> ingredient) throws IllegalArgumentException {
+        checkState(!this.aisle.isEmpty(), "aisle must be set before setting aisle symbols");
+
+        if (ingredient != null) {
+            this.ingredientMap.put(symbol, ingredient);
+        } else {
+            this.ingredientMap.remove(symbol);
+        }
+
+        return this;
+    }
+
+    @SuppressWarnings("ConstantConditions")
+    @Override
+    public ShapedCraftingRecipe.Builder where(char symbol, ItemStackSnapshot ingredient) throws IllegalArgumentException {
+        return where(symbol, ingredient != null && ingredient != ItemStackSnapshot.NONE
+                ? new MatchCraftingVanillaItemStack(ingredient) : null);
+    }
+
+    @Override
+    public ShapedCraftingRecipe.Builder result(ItemStackSnapshot result) {
+        Preconditions.checkNotNull(result, "result");
+
+        this.result = result;
+
+        return this;
+    }
+
+    @Override
+    public ShapedCraftingRecipe build() {
+        checkState(!this.aisle.isEmpty(), "aisle has not been set");
+        checkState(!this.ingredientMap.isEmpty(), "no ingredients set");
+        checkState(this.result != ItemStackSnapshot.NONE, "no result set");
+
+        ImmutableTable.Builder<Integer, Integer, Predicate<ItemStackSnapshot>> tableBuilder = ImmutableTable.builder();
+        Iterator<String> aisleIterator = this.aisle.iterator();
+        String aisleRow = aisleIterator.next();
+        int width = aisleRow.length();
+        int height = 0;
+
+        checkState(width > 0, "The aisle cannot be empty.");
+
+        do {
```

Comment:
This is a very weird loop. I'm confused as to why it can't just be a for-loop with the iterator...